File: lib/app/modules/statistical/statistical_view/components/ring_progress_painter.dart
import 'dart:math' as math;

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

class RingProgressPainter extends CustomPainter {
  final double progress;
  final Color color;
  final Color trackColor;
  final double strokeWidth;
  final List<Color>? gradientColors;

  const RingProgressPainter({
    required this.progress,
    required this.color,
    required this.trackColor,
    required this.strokeWidth,
    this.gradientColors,
  });

  static const double _startAngle = math.pi / 2;
  static const double _totalSweep = 2 * math.pi;

  @override
  void paint(Canvas canvas, Size size) {
    final rect = (Offset.zero & size).deflate(strokeWidth / 2);

    final track = Paint()
      ..color = trackColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round;
    canvas.drawArc(rect, _startAngle, _totalSweep, false, track);

    if (progress <= 0) {
      return;
    }

    final sweep = _totalSweep * progress.clamp(0.0, 1.0);

    final arc = Paint()
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth
      ..strokeCap = StrokeCap.round;

    final colors = gradientColors;
    if (colors != null && colors.length >= 2) {
      final capAngle = strokeWidth / rect.width;
      arc.shader = SweepGradient(
        startAngle: 0,
        endAngle: math.max(sweep + capAngle, 0.01),
        colors: colors,
        transform: GradientRotation(_startAngle - capAngle),
      ).createShader(rect);
    } else {
      arc.color = color;
    }

    canvas.drawArc(rect, _startAngle, sweep, false, arc);
  }

  @override
  bool shouldRepaint(RingProgressPainter oldDelegate) =>
      oldDelegate.progress != progress ||
      oldDelegate.color != color ||
      oldDelegate.trackColor != trackColor ||
      oldDelegate.strokeWidth != strokeWidth ||
      !listEquals(oldDelegate.gradientColors, gradientColors);
}

File: android/app/src/main/kotlin/com/example/project/widget/RingProgressPainter.kt
package com.example.project.widget

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.SweepGradient
import androidx.annotation.ColorInt
import kotlin.math.max

object RingProgressPainter {

    private const val START_ANGLE = 90f
    private const val TOTAL_SWEEP = 360f

    fun render(
        context: Context,
        sizeDp: Float,
        strokeWidthDp: Float,
        progress: Float,
        @ColorInt color: Int,
        @ColorInt trackColor: Int,
        gradientColors: IntArray? = null,
    ): Bitmap {
        val density = context.resources.displayMetrics.density
        val size = (sizeDp * density).toInt().coerceAtLeast(1)
        val stroke = strokeWidthDp * density

        val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)

        val inset = stroke / 2f
        val rect = RectF(inset, inset, size - inset, size - inset)

        val track = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            this.color = trackColor
            style = Paint.Style.STROKE
            strokeWidth = stroke
            strokeCap = Paint.Cap.ROUND
        }
        canvas.drawArc(rect, START_ANGLE, TOTAL_SWEEP, false, track)

        if (progress <= 0f) return bitmap

        val sweep = TOTAL_SWEEP * progress.coerceIn(0f, 1f)

        val arc = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE
            strokeWidth = stroke
            strokeCap = Paint.Cap.ROUND
        }

        arc.applyStroke(rect, sweep, stroke, color, gradientColors)

        canvas.drawArc(rect, START_ANGLE, sweep, false, arc)

        return bitmap
    }

    private fun Paint.applyStroke(
        rect: RectF,
        sweep: Float,
        stroke: Float,
        @ColorInt color: Int,
        gradientColors: IntArray?,
    ) {
        if (gradientColors == null || gradientColors.size < 2) {
            this.color = color
            return
        }

        val capAngle = Math.toDegrees((stroke / rect.width()).toDouble()).toFloat()
        val span = max(sweep + capAngle, 0.01f)

        val frac = (span / TOTAL_SWEEP).coerceIn(0f, 1f)
        val last = gradientColors.size - 1
        val positions = FloatArray(gradientColors.size) { i -> i.toFloat() / last * frac }

        shader = SweepGradient(rect.centerX(), rect.centerY(), gradientColors, positions).apply {
            setLocalMatrix(
                Matrix().apply {
                    setRotate(START_ANGLE - capAngle, rect.centerX(), rect.centerY())
                },
            )
        }
    }
}
